qiskit/optimization/problems/quadratic_program.py

```diff
@@ -152,36 +153,139 @@ def _add_variable(self,
                       upperbound: Union[float, int] = INFINITY,
                       vartype: VarType = VarType.CONTINUOUS,
                       name: Optional[str] = None) -> Variable:
-        """Checks whether a variable name is already taken and adds the variable to list and index
-        if not.
+        return self._add_variables(lowerbound, upperbound, vartype, name)[1][0]
+
+    def _add_variables(self,
+                       lowerbound: Union[float, int] = 0,
+                       upperbound: Union[float, int] = INFINITY,
+                       vartype: VarType = VarType.CONTINUOUS,
+                       name: Optional[str] = None,
+                       key_format: Optional[str] = None,
+                       keys: Union[int, Sequence] = 1) -> Tuple[List[str], List[Variable]]:
+        if isinstance(keys, int) and keys < 1:
+            raise QiskitOptimizationError(
+                "Cannot create non-positive number of variables: {}".format(keys))
+        if not name and key_format:
+            name = 'x'
+        if name and not key_format:
+            if isinstance(keys, int) and keys > 1 or isinstance(keys, Sequence) and len(keys) > 0:
+                key_format = '{}'
+            else:
+                key_format = ''
+                if name in self._variables_index:
+                    raise QiskitOptimizationError("Variable name already exists: {}".format(name))
+        if not name and not key_format:
+            name = 'x'
+            key_format = '{}'
+        if '{{}}' in key_format:
+            raise QiskitOptimizationError(
+                "Formatter cannot contain nested substitutions: {}".format(key_format))
+        substitution_count = key_format.count('{}')
+        if substitution_count > 1:
+            raise QiskitOptimizationError(
+                "Formatter cannot contain more than one substitution: {}".format(key_format))
+
+        names = []
+        variables = []
+        if isinstance(keys, Sequence):
+            for key in keys:
+                indexed_name = name + key_format.format(str(key))
```

Comment:
I don't think you need to convert `key` into `str` explicitly. `key_format.format` is expected to do it.

Reply:
For example, `qp.binary_var_dict(name='x', key_format='{:02d}', keys=10)` works well, but it does not work with range as follows.
```python
qp.binary_var_dict(name='x', key_format='{:02d}', keys=range(10))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-38dca7db3ac8> in <module>
----> 1 qp.binary_var_dict(name='x', key_format='{:02d}', keys=range(10))

~/tasks/0_2020/qiskit/aqua/qiskit/optimization/problems/quadratic_program.py in binary_var_dict(self, name, key_format, keys)
    401                                      nested substitution.
    402         """
--> 403         return self._var_dict(0, 1, Variable.Type.BINARY, name, key_format, keys)
    404 
    405     def binary_var_list(self,

~/tasks/0_2020/qiskit/aqua/qiskit/optimization/problems/quadratic_program.py in _var_dict(self, lowerbound, upperbound, vartype, name, key_format, keys)
    251         """
    252         return dict(
--> 253             zip(*self._add_variables(lowerbound, upperbound, vartype, name, key_format, keys)))
    254 
    255     def _var_list(self,

~/tasks/0_2020/qiskit/aqua/qiskit/optimization/problems/quadratic_program.py in _add_variables(self, lowerbound, upperbound, vartype, name, key_format, keys)
    190         if isinstance(keys, Sequence):
    191             for key in keys:
--> 192                 indexed_name = name + key_format.format(str(key))
    193                 if indexed_name in self._variables_index:
    194                     raise QiskitOptimizationError(

ValueError: Unknown format code 'd' for object of type 'str'
```